### PR TITLE
Make target tensor available to processors

### LINKED ISSUE #108: Make target available during training
### Problem / Goal

Some models (e.g. diffusion) need the target tensor to be available during training.

### What does good look like?

- [ ] Target tensor is available during training

### Related context

_No response_

### PR DESCRIPTION
**Pull request by jemrobinson**
Make the `target` tensor available to instances of `BaseProcessor` during the encode-process-decode workflow.

Child classes will need to override `rollout` which has the following type signature `rollout(self, x: TensorNTCHW, y: TensorNTCHW | None = None) -> TensorNTCHW`. The inputs are:

- `x` input tensor in latent space `(batch_size, n_history_steps, n_latent_channels_total, latent_height, latent_width)`
- `y`: target tensor in latent space `(batch_size, n_forecast_steps, n_latent_channels_total, latent_height, latent_width)`
- output: predicted tensor in latent space `(batch_size, n_forecast_steps, n_latent_channels_total, latent_height, latent_width)`

:warning: The same `rollout` method is used for test/train/validate/predict. For `train` and `validate`, `y` will be provided, but for `test` and `predict` it will be `None`. Code that wants to use this tensor should check for `y is None` before doing so.

## Other changes
Swapped the definition of `forward` and `rollout`. `forward` now does `TensorNCHW -> TensorNCHW` for a single timestep and the default implementation of `rollout` is to call `forward` N times. This should better correspond to the logical meanings of `forward` and `rollout`.

Closes #108

### CODE REVIEW

**Comment by marianovitasari20** [comment]
Hi @jemrobinson, quick clarification: I noticed you mentioned that `y` will be `None` for everything except training. But for validation (and often test), don’t we usually still need `y` so we can compute metrics against the ground truth? I thought only predict would typically have no `y`. Am I missing something about how our pipeline is set up?

**Review by louisavz — COMMENTED** [review]
After further discussion with @jemrobinson, the pipeline already handles training and validation loop. I don't know if the additional `y` is necessary without further reviewing the diffusion code, but perhaps best to leave it as is, as long as it doesn't break PR #85 that is in the process of being reviewed and merged. Perhaps better yet, don't add the `y` here, but utilise the pipeline to call the training step. Would love to hear what others think.

**Comment by marianovitasari20** [comment]
Just to clarify, generative models like DDPM don’t work the same way as common regression-style models (e.g., U-Net). For training, DDPM necessarily requires *y*, since the objective is defined by corrupting the clean target *y* with noise and then learning to denoise it. Without *y*, the model cannot be trained.

**Comment by marianovitasari20** [comment]
For example, with the current pipeline, training and validation are forced to compare `y` and `y_hat`. That works for common regression-style models, but for generative models (e.g. DDPM), it’s different:

* **Training:** the loss is computed on predicted noise (`ε_pred`) or v-prediction (`v_pred`) vs the corresponding target derived from `y`, not `y` vs `y_hat`. Changing between ε and v already changes the variables used for the loss.
* **Validation:** we evaluate forecast skill using metrics on the generated predictions against `y`, but the training loss itself is still `v_pred`-based.

This is exactly why we need overrides for each model’s logic, the central pipeline can’t assume a single loss formulation or evaluation approach.

File: ice_station_zebra/models/processors/base_processor.py [hunk 1]
```diff
@@ -19,54 +19,55 @@ def __init__(
...
 
-        If you want to handle the NTCHW tensors directly, simply override this method in
-        your child class.
+        If you want to handle the NTCHW tensors directly, or to use the target tensor for
+        model training simply override this method in your child class.
 
         Args:
-            x: TensorNTCHW with (batch_size, n_history_steps, n_latent_channels_total, latent_height, latent_width)
+            x: Input TensorNTCHW with (batch_size, n_history_steps, n_latent_channels_total, latent_height, latent_width)
+            y: during training: Target TensorNTCHW with (batch_size, n_forecast_steps, n_latent_channels_total, latent_height, latent_width)
```
[the remark this answers is not in the document]
Comment (marianovitasari20) [reply ice_station_zebra/models/processors/base_processor.py:54]: > They both appear to treat validation separated from the training.

They are always separated. That’s why PyTorch Lightning has separate hooks for `training`, `validation`, `test`, and `predict`, each stage is encapsulated to avoid mixing logic and to handle losses and metrics appropriately for that stage.

[the remark this answers is not in the document]
Comment (marianovitasari20) [reply ice_station_zebra/models/processors/base_processor.py:54]: > After further discussion with @jemrobinson, the pipeline already handles training and validation loop.

Yes, but with this centralised loop, we’re forced to use the same loss function for all models, which isn’t ideal. I explained in more detail in another comment why this can be problematic.

[the remark this answers is not in the document]
Comment (marianovitasari20) [reply ice_station_zebra/models/processors/base_processor.py:54]: I’m not sure what you mean by:
> unless we are pulling the validation set from the training set in one of  the discussions 

 Are you referring to training/validation data separation? The training and validation logic is always separate. PyTorch handles this automatically,

**Comment by marianovitasari20** [comment]
Happy to elaborate if that would be helpful. Please do let me know if you need further discussion about this.

File: ice_station_zebra/models/processors/base_processor.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jemrobinson) [reply ice_station_zebra/models/processors/base_processor.py:54]: y added to validation in 72b58bf
